bin/main.rb

```diff
@@ -10,61 +11,54 @@
 
 puts 'Enter your name Player 1: '.colorize(:green)
 player1 = gets.chomp
+player1 = Player.new(player1, 'X')
 
 puts 'Enter your name Player 2: '.colorize(:green)
 player2 = gets.chomp
+player2 = Player.new(player2, 'O')
 
-board = %w[1 2 3 4 5 6 7 8 9]
+grid = Grid.new
 
-game_on = true
-# While game end
-winner = nil
-while game_on
+until grid.winner
 
-  def return_board(board)
-    puts "| #{board[0]} | #{board[1]} | #{board[2]} |".colorize(:light_blue)
-    puts '--------------'.colorize(:light_blue)
-    puts "| #{board[3]} | #{board[4]} | #{board[5]} |".colorize(:light_blue)
-    puts '--------------'.colorize(:light_blue)
-    puts "| #{board[6]} | #{board[7]} | #{board[8]} |".colorize(:light_blue)
-    puts "\n"
-  end
-
-  puts "#{player1} pick a number from 1 to 9: ".colorize(:green)
-  # show the board without selection
-  return_board(board)
-  input1 = gets.chomp
-  input1 = input1.to_i
+  puts grid.board_display
 
-  board[input1.to_i - 1] = 'X'
-  puts "#{player1} has selected the #{input1} position".colorize(:green)
-  return_board(board)
-  # show the board with selection
+  puts "#{player1.name} pick a number from 1 to 9: ".colorize(:green)
 
-  puts "#{player2} pick a number from 1 to 9: ".colorize(:green)
-  # show the board without selection
-  return_board(board)
-  input2 = gets.chomp
-  if board[input2.to_i - 1] == 'X'
-    print 'Sorry this move is not valid, '
-    p 'please select another place'
-  else
-    board[input2.to_i - 1] = 'O'
-    puts "#{player2} has selected the #{input2} position".colorize(:green)
-    return_board(board)
+  puts grid.board_display
+  loop do
+    input1 = gets.chomp.to_i
+    if grid.check_valid?(input1 - 1)
+      grid.update_board(player1, input1 - 1)
+      break
+    else
+      puts "#{input1} is already occupied"
+      puts "#{player1.name} pick a number from 1 to 9: ".colorize(:green)
+    end
   end
 
-  if board.count('X') == 3
-    winner ||= player1
-  elsif board.count('O') == 3
-    winner ||= player2
-  elsif board.count('X') == 4 && board.count('O') == 4
-    winner ||= 'draw'
-  end
+  puts "#{player1.name} has selected the #{input1} position".colorize(:green)
```

Comment:
- [ ] Your code breaks when it gets to line 40, and that is because you are trying to access a variable that is not available in that scope. You declared the input1 variable inside of a loop and you are trying to access that variable outside of that loop. So you might need to declare the variable outside of the loop so that you can access it both inside and outside of the loop.